### PR TITLE
feat: iavl v2

### PR DESCRIPTION
**Pull request by kocubinski**
See also: https://hackmd.io/rXq7coSyTFmhWIa4mifBYQ

For a complete end to end test, this version of code is running on an osmosis node syncing on mainnet, some metrics are viewable at [this Grafana](http://osmosis-iavl-v2.kocubinski.com:3000/d/fa4937cf-92ec-47b8-90c9-fc3d32957ebc/osmosis-bench?orgId=1&refresh=5s&from=now-30m&to=now).  Sync started from block `13946655` after an export of `application.db` to an iavl v2 SQLite database with the code in `./migrate` in this PR.

Sync rates as fast as ~10 blocks per second are observed, most of the spikiness in commit rate is spent waiting on blocks from consensus, but I don't have a way to show this in telemetry.  AFAIK this is about 2x from iavl v1 with osmosis,  I don't see evidence that the osmosis build is I/O bound on disk either as benchmarks put max throughput 10-20x iavl v1.

This configuration is using 45G of disk storage, down from 177G.
Checkpoints are taken every 5k blocks, with a WAL maintained on each commit.
Pruning occurs every 900 blocks with the last 5k blocks kept, therefore the last checkpoint always contains a fully materialized tree.

### CODE REVIEW

**Comment by tac0turtle** [comment]
is this ready for review?

File: v2/export.go
Comment (tac0turtle) [inline v2/export.go:19]: go docs

File: v2/internal/encoding.go
Comment (tac0turtle) [inline v2/internal/encoding.go:31]:
```suggestion
// DecodeBytes decodes a varint length-prefixed byte slice, returning it along with the number
```

**Comment by cool-develope** [comment]
Rollback feature is missed (DeleteVersionsFrom in v1). 
@kocubinski, I think MultiTree has the potential advantages in terms of deferred pruning and also would be easy to manage the flow.

**Comment by cool-develope** [comment]
how do we test v2 ?

File: v2/cmd/gen/gen.go
Comment (odeke-em) [inline v2/cmd/gen/gen.go:63]: Missing defer iter.Close()

File: v2/cmd/gen/gen.go
Comment (odeke-em) [inline v2/cmd/gen/gen.go:74]: Please add a defer to close the channel as well as wait on the wait group per

```go
defer func() {
  close(ch)
  wg.Wait()
}()
```

File: v2/cmd/gen/gen.go
Comment (odeke-em) [inline v2/cmd/gen/gen.go:78]: It's not recommended to invoke log.Fatal in a go routine, it doesn't give others time to cleanup.

File: v2/cmd/gen/gen.go
Comment (odeke-em) [inline v2/cmd/gen/gen.go:170]: We need to catch the end of the iteration, but this code just returns on any generic error

File: v2/cmd/gen/gen.go
Comment (odeke-em) [inline v2/cmd/gen/gen.go:178]: How does this code differentiate between the end of the iterator and another error? I don't see any special casing to break out of the loop

File: v2/cmd/latest.go
Comment (odeke-em) [inline v2/cmd/latest.go:33]: Missing sql.Close() on ending the function.

File: v2/cmd/scan/scan.go
Comment (odeke-em) [inline v2/cmd/scan/scan.go:28]: Missing defer f.Close()

File: v2/cmd/scan/scan.go
Comment (odeke-em) [inline v2/cmd/scan/scan.go:68]: Without f.Close() this code is undeterministic and might not have flushed data but even more on some systems, trying to delete an open file can be refused by the security policy.

Comment (odeke-em) [reply v2/cmd/scan/scan.go:68]: Also you can simply invoke a return instead of special casing err != nil, vs nil error

File: v2/cmd/snapshot/snapshot.go
Comment (odeke-em) [inline v2/cmd/snapshot/snapshot.go:42]: This handle is never closed on unrelated errors. Please close it after use perhaps with a defer, or in each error flow clause.

File: v2/export.go
Comment (odeke-em) [inline v2/export.go:22]: To speed this code up, please make these buffered channels with some capacity by means of a heuristic, so that traversal won't be constrained and blocked by the speed of reading out the nodes

**Review by odeke-em — COMMENTED** [review]
Thank you @kocubinski, I just saw this change and here is a first pass review.

**Review by tac0turtle — APPROVED** [review]

**Comment by tac0turtle** [comment]
gonna merge this as john is working iavlv2 cleann

**Pull request merged**
